### PR TITLE
Move default socket path into XDG_RUNTIME_DIR if defined

### PR DESCRIPTION
**Pull request by tristan957**
> $XDG_RUNTIME_DIR defines the base directory relative to which
user-specific non-essential runtime files and other file objects (such as sockets, named pipes, ...) should be stored. The directory MUST be owned by the user, and he MUST be the only one having read and write access to it. Its Unix access mode MUST be 0700.

> The lifetime of the directory MUST be bound to the user being logged
in. It MUST be created when the user first logs in and if the user fully logs out the directory MUST be removed. If the user logs in more than once he should get pointed to the same directory, and it is mandatory that the directory continues to exist from his first login to his last logout on the system, and not removed in between. Files in the directory MUST not survive reboot or a full logout/login cycle.

> The directory MUST be on a local file system and not shared with any
other system. The directory MUST by fully-featured by the standards of the operating system. More specifically, on Unix-like operating systems AF_UNIX sockets, symbolic links, hard links, proper permissions, file locking, sparse files, memory mapping, file change notifications, a reliable hard link count must be supported, and no restrictions on the file name character set should be imposed. Files in this directory MAY be subjected to periodic clean-up. To ensure that your files are not removed, they should have their access time timestamp modified at least once every 6 hours of monotonic time or the 'sticky' bit should be set on the file.

> If $XDG_RUNTIME_DIR is not set applications should fall back to a
replacement directory with similar capabilities and print a warning message. Applications should use this directory for communication and synchronization purposes and should not place larger files in it, since it might reside in runtime memory and cannot necessarily be swapped out to disk.

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

According to the XDG Base Directory Specification, this is where our socket path should exist by default. In the event it doesn't exist, fallback to /tmp as the parent directory.

Signed-off-by: Tristan Partin <tpartin@micron.com>

### CODE REVIEW

**Comment by alexttx** [comment]
> The lifetime of the directory MUST be bound to the user being logged in

What happens if a user logs in, starts HSE with nohup in the background, and logs out.  Does the XGD_RUNTIME_DIR remain?

**Comment by tristan957** [comment]
The directory sticks around.

File: lib/config/hse_gparams.c [hunk 1]
```diff
@@ -103,15 +104,21 @@ logging_destination_jsonify(const struct param_spec *const ps, const void *const
...
-
+    const char *dir;
     HSE_MAYBE_UNUSED int n;
 
-    n = snprintf(value, sizeof(hse_gparams.gp_socket.path), "/tmp/hse-%d.sock", getpid());
+    INVARIANT(ps);
+    INVARIANT(value);
+
+    dir = getenv("XDG_RUNTIME_DIR");
+    if (!dir || strlen(dir) == 0)
```
Comment (tristan957) [inline lib/config/hse_gparams.c:117]: *What do you guys thing of the strlen() check here? Unnecessary?

**Review by alexttx — APPROVED** [review]

File: lib/config/hse_gparams.c (hunk 1 above)
Comment (alexttx) [reply lib/config/hse_gparams.c:117]: From https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

> All paths set in these environment variables must be absolute. If an implementation encounters a relative path in any of these variables it should consider the path invalid and ignore it.

You  could use:   `if (!dir || *dir != '/')`

Comment (tristan957) [reply lib/config/hse_gparams.c:117]: Great catch.

**Review by nabeelmmd — APPROVED** [review]

**Review by beckerg — APPROVED** [review]

**Pull request merged**